Code/Framework/AzCore/AzCore/Math/VectorN.h

```diff
@@ -0,0 +1,149 @@
+        //! Comparison functions, not implemented as operators since that would probably be a little dangerous.
+        //! These functions return true only if all components pass the comparison test.
+        //! @{
+        bool IsLessThan(const VectorN& v) const;
+        bool IsLessEqualThan(const VectorN& v) const;
+        bool IsGreaterThan(const VectorN& v) const;
+        bool IsGreaterEqualThan(const VectorN& v) const;
+        //! @}
+
+        //! Floor/Ceil/Round functions, operate on each component individually, result will be stored in output.
+        //! @{
+        VectorN GetFloor() const;
+        VectorN GetCeil() const;
+        VectorN GetRound() const; // Ties to even (banker's rounding)
+        //! @}
+
+        //! Min/Max functions, operate on each component individually, result will be stored in output.
+        //! @{
+        VectorN GetMin(const VectorN& v) const;
+        VectorN GetMax(const VectorN& v) const;
+        VectorN GetClamp(const VectorN& min, const VectorN& max) const;
+        //! @}
+
+        //! Returns squared length of the vector.
+        float GetLengthSq() const;
+
+        //! Returns length of the vector, full accuracy.
+        float GetLength() const;
+
+        //! Returns normalized vector, full accuracy.
+        VectorN GetNormalized() const;
+
+        //! Normalizes the vector in-place, full accuracy.
+        void Normalize();
+
+        //! Returns a new VectorN containing the absolute value of all elements in the source VectorN.
+        VectorN GetAbs() const;
+
+        //! Returns a new VectorN containing the square of all elements in the source VectorN.
+        VectorN GetSquare() const;
+
+        //! Returns the dot product of two vectors of equal dimension.
+        float Dot(const VectorN& rhs) const;
+
+        VectorN& operator+=(const VectorN& rhs);
+        VectorN& operator-=(const VectorN& rhs);
+        VectorN& operator*=(const VectorN& rhs); //! Hadamard product, not dot product.
+        VectorN& operator/=(const VectorN& rhs);
+
+        VectorN& operator+=(float sum);
+        VectorN& operator-=(float difference);
+        VectorN& operator*=(float multiplier);
+        VectorN& operator/=(float divisor);
+
+        VectorN operator-() const;
+        VectorN operator+(const VectorN& rhs) const;
+        VectorN operator-(const VectorN& rhs) const;
+        VectorN operator*(const VectorN& rhs) const; //! Hadamard product, not dot product.
+        VectorN operator/(const VectorN& rhs) const;
+
+        VectorN operator*(float multiplier) const;
+        VectorN operator/(float divisor) const;
+
+        //! Returns the raw Vector4's that represent this vector instance.
+        const AZStd::vector<Vector4>& GetVectorValues() const;
+        AZStd::vector<Vector4>& GetVectorValues();
+
+        //! Zeros out unused components of the last vector element
+        void FixLastVectorElement();
+
+    private:
+
+        const AZStd::size_t m_numElements = 0;
```

Comment:
Because of the const member, the implicit `move assignment operator` is deleted because it assumes a copy has to be made.
It is better of for both this class and `MatrixMxN` to just have the number of element members be non-const.